File: bindings/python/examples/gpioset.py
#!/usr/bin/env python3
# SPDX-License-Identifier: GPL-2.0-or-later
# SPDX-FileCopyrightText: 2022 Bartosz Golaszewski <brgl@bgdev.pl>

"""Simplified reimplementation of the gpioset tool in Python."""

import gpiod
import sys

from gpiod.line import Direction, Value

if __name__ == "__main__":
    if len(sys.argv) < 3:
        raise TypeError(
            "usage: gpioset.py <gpiochip> <offset1>=<value1> <offset2>=<value2> ..."
        )

    path = sys.argv[1]

    def parse_value(arg):
        x, y = arg.split("=")
        return (x, Value(int(y)))

    def make_settings(val):
        return gpiod.LineSettings(direction=Direction.OUTPUT, output_value=val)

    lvs = [parse_value(arg) for arg in sys.argv[2:]]
    config = dict([(l, make_settings(v)) for (l, v) in lvs])

    request = gpiod.request_lines(
        path,
        consumer="gpioset.py",
        config=config,
    )

    input()

File: bindings/rust/libgpiod/examples/gpioset.rs
// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
// SPDX-FileCopyrightText: 2022 Linaro Ltd.
// SPDX-FileCopyrightText: 2022 Viresh Kumar <viresh.kumar@linaro.org>
//
// Simplified Rust implementation of the gpioset tool.

use std::env;
use std::io::{stdin, Read};

use libgpiod::{
    chip::Chip,
    line::{self, Direction, Offset, SettingVal, Value},
    request, Error, Result,
};

fn usage(name: &str) {
    println!("Usage: {} <chip> <line_offset0>=<value0> ...", name);
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 3 {
        usage(&args[0]);
        return Err(Error::InvalidArguments);
    }

    let mut lconfig = line::Config::new()?;

    for arg in &args[2..] {
        let pair: Vec<&str> = arg.split('=').collect();
        if pair.len() != 2 {
            usage(&args[0]);
            return Err(Error::InvalidArguments);
        }

        let offset = pair[0]
            .parse::<Offset>()
            .map_err(|_| Error::InvalidArguments)?;
        let value = pair[1]
            .parse::<i32>()
            .map_err(|_| Error::InvalidArguments)?;

        let mut lsettings = line::Settings::new()?;
        lsettings.set_prop(&[
            SettingVal::Direction(Direction::Output),
            SettingVal::OutputValue(Value::new(value)?),
        ])?;
        lconfig.add_line_settings(&[offset], lsettings)?;
    }

    let path = format!("/dev/gpiochip{}", args[1]);
    let chip = Chip::open(&path)?;

    let mut rconfig = request::Config::new()?;
    rconfig.set_consumer(&args[0])?;

    chip.request_lines(Some(&rconfig), &lconfig)?;

    // Wait for keypress, let user verify line status.
    stdin().read_exact(&mut [0u8]).unwrap();

    Ok(())
}
